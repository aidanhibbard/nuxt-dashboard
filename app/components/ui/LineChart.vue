<script setup lang="ts">
import { computed } from 'vue'
import {
  Chart as ChartJS,
  CategoryScale,
  LinearScale,
  PointElement,
  LineElement,
  Title,
  Tooltip,
  Legend,
} from 'chart.js'
import { Line } from 'vue-chartjs'

ChartJS.register(CategoryScale, LinearScale, PointElement, LineElement, Title, Tooltip, Legend)

type InputDataset = {
  label: string
  data: readonly number[]
  backgroundColor?: string | readonly string[]
  borderColor?: string | readonly string[]
}

type MutableDataset = {
  label: string
  data: number[]
  backgroundColor?: string | string[]
  borderColor?: string | string[]
}

const props = defineProps<{ labels: readonly string[], datasets: readonly InputDataset[], height?: number }>()

const options = {
  responsive: true,
  maintainAspectRatio: false,
  plugins: { legend: { display: true } },
}

const chartData = computed<{ labels: string[], datasets: MutableDataset[] }>(() => ({
  labels: [...props.labels],
  datasets: props.datasets.map(d => ({
    label: d.label,
    data: [...d.data],
    backgroundColor: Array.isArray(d.backgroundColor)
      ? [...d.backgroundColor] as string[]
      : (d.backgroundColor as string | undefined),
    borderColor: Array.isArray(d.borderColor)
      ? [...d.borderColor] as string[]
      : (d.borderColor as string | undefined),
  })),
}))
</script>

<template>
  <div :style="{ height: `${props.height ?? 300}px` }">
    <Line
      :data="chartData"
      :options="options"
    />
  </div>
</template>
