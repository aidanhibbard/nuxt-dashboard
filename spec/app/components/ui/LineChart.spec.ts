import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import LineChart from '@/components/ui/LineChart.vue'

describe('components/ui/LineChart.vue', () => {
  it('renders chart container', () => {
    const wrapper = mount(LineChart, {
      props: {
        labels: ['Jan', 'Feb'],
        datasets: [{ label: 'A', data: [1, 2] }],
        height: 200,
      },
    })
    expect(wrapper.find('.chart-line').exists()).toBe(true)
  })

  it('matches snapshot', () => {
    const wrapper = mount(LineChart, {
      props: {
        labels: ['Jan', 'Feb', 'Mar'],
        datasets: [{ label: 'A', data: [1, 3, 2] }],
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})


