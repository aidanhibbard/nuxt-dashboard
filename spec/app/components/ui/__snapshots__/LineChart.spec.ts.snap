// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components/ui/LineChart.vue > matches snapshot 1`] = `
"<div style="height: 300px;">
  <div class="chart-line" data="[object Object]" options="[object Object]"></div>
</div>"
`;
